### PR TITLE
8339252: Add method float16Value() to BigInteger and BigDecimal

### PR DESCRIPTION
**Pull request by rgiulietti**
This patch adds `float16Value()` to `BigInteger` and `BigDecimal`, as well as `Float16.valueOf(BigDecimal)`.

---------
### Progress
- [x] Change must not contain extraneous whitespace

### Issues
 * [JDK-8339252](https://bugs.openjdk.org/browse/JDK-8339252): Add method float16Value() to BigInteger and BigDecimal (**Enhancement** - P4)
 * [JDK-8337761](https://bugs.openjdk.org/browse/JDK-8337761): Add BigDecimal to Float16 conversion (**Enhancement** - P4)

### Reviewers
 * [Joe Darcy](https://openjdk.org/census#darcy) (@jddarcy - no project role)
 * [Jatin Bhateja](https://openjdk.org/census#jbhateja) (@jatin-bhateja - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/valhalla.git pull/1228/head:pull/1228` \
`$ git checkout pull/1228`

Update a local copy of the PR: \
`$ git checkout pull/1228` \
`$ git pull https://git.openjdk.org/valhalla.git pull/1228/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1228`

View PR using the GUI difftool: \
`$ git pr show -t 1228`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/valhalla/pull/1228.diff">https://git.openjdk.org/valhalla/pull/1228.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/valhalla/pull/1228#issuecomment-2321158986)

### CODE REVIEW

**Comment by rgiulietti** [comment]
I erroneously had this PR on the `lworld` branch, then moved it to the `lworld+fp16` branch, where it belongs.

However, the webrevs now show the diffs w.r.t. `lworld`, which is quite confusing.
In addition, the "Files changed" tab shows diffs for `Float16.valueOf(long)` that should not appear in `lworld+fp16`.

Short of closing this PR and open it again in the right branch (but I'm not sure this works without conflicts), I'm clueless.
So I'm asking for advice.

File: src/java.base/share/classes/java/math/BigDecimal.java [hunk 1]
```diff
@@ -3776,6 +3780,98 @@ public byte byteValueExact() {
...
+        } else {
+            m = w;
+            n = pow10.shiftLeft(ql);
+        }
+        BigInteger[] qr = m.divideAndRemainder(n);
+        /*
+         * We have
+         *      2^12 = 2^{P+1} <= i < 2^{P+5} = 2^16
+         * Contrary to the double and float cases, here we cannot simply
+         * declare i as short, because P + 5 < Short.SIZE fails to hold.
```
Comment (jatin-bhateja) [inline src/java.base/share/classes/java/math/BigDecimal.java:3857]: Comment is not very clear here, can you please elaborate on the significance of augend '5' here.

File: test/jdk/java/math/BigInteger/PrimitiveConversionTests.java [hunk 2]
```diff
@@ -98,9 +98,45 @@ public static int testFloatValue() {
...
+        int failures = 0;
+        for (int i = 0; i <= high; ++i) {
+            BigInteger big = BigInteger.valueOf(i);
+            failures = checkFloat16(Float16.valueOf(i), big.float16Value(), big, failures);
+
+            big = BigInteger.valueOf(-i);
+            failures = checkFloat16(Float16.valueOf(-i), big.float16Value(), big, failures);
+        }
+
+        BigInteger large = ONE.shiftLeft(1_000);
```
[the remark this answers is not in the document]
Comment (rgiulietti) [reply test/jdk/java/math/BigInteger/PrimitiveConversionTests.java:117]: All integers between `-(MAX_VALUE + ULP(MAX_VALUE))` and `(MAX_VALUE + ULP(MAX_VALUE))` are tested in the loop, including `MAX_VALUE + 0.5 * ULP(MAX_VALUE)`.

But we also need a test for really big integers.

File: src/java.base/share/classes/java/math/BigDecimal.java (hunk 1 above)
Comment (rgiulietti) [reply src/java.base/share/classes/java/math/BigDecimal.java:3857]: The full documentation of the algorithm is in extensive comments in the `doubleValue()` and `fullDoubleValue()` methods. Here, we only apply that logic as much as possible, adapting it for the `Float16` case.

In particular, see the discussion starting on L.4158

Comment (rgiulietti) [reply src/java.base/share/classes/java/math/BigDecimal.java]: I hope I made the comment a bit clearer.

File: src/java.base/share/classes/java/lang/Float16.java [hunk 3]
```diff
@@ -282,11 +283,11 @@ public static Float16 valueOf(int value) {
     * @param  value a {@code long} value.
     */
     public static Float16 valueOf(long value) {
-        if (value < -65_504L) {
+        if (value <= -65_520L) {  // -(Float16.MAX_VALUE + Float16.ulp(Float16.MAX_VALUE) / 2)
```
Comment (jddarcy) [inline src/java.base/share/classes/java/lang/Float16.java:286]: I assume you'll sync in / merge this change from the branch.

File: src/java.base/share/classes/java/math/BigDecimal.java [hunk 4]
```diff
@@ -3776,6 +3780,99 @@ public byte byteValueExact() {
...
+        BigInteger w = unscaledValue().abs();
+        long qb = w.bitLength() - (long) Math.ceil(scale * L);
+        if (qb < Q_MIN_F16 - 2) {  // qb < -26
+            return Float16.valueOf(signum() * 0.0f);
+        }
+        if (qb > Q_MAX_F16 + P_F16 + 1) {  // qb > 17
+            return Float16.valueOf(signum() * Float.POSITIVE_INFINITY);
+        }
+        if (scale < 0) {
+            return Float16.valueOf(signum() * w.multiply(bigTenToThe(-scale)).floatValue());
```
Comment (jddarcy) [inline src/java.base/share/classes/java/math/BigDecimal.java:3837]: Are the calls to `floatValue()` for the `scale < 0` and `scale == 0` cases meant to be calls to `float16Value()`?

If not, the different coding pattern from the analagous float and double methods should be documented in a comment.

File: src/java.base/share/classes/java/lang/Float16.java (hunk 3 above)
Comment (rgiulietti) [reply src/java.base/share/classes/java/lang/Float16.java:286]: I think this is a remnant due to my previous error to associate the PR to branch `lworld` before moving it to `lworld+fp16`.

File: src/java.base/share/classes/java/math/BigDecimal.java (hunk 4 above)
Comment (rgiulietti) [reply src/java.base/share/classes/java/math/BigDecimal.java]: Improved code to mirror analogous code in `fullDoubleValue()`.

File: src/java.base/share/classes/java/lang/Float16.java [hunk 5]
```diff
@@ -419,6 +420,20 @@ public static Float16 valueOf(String s) throws NumberFormatException {
...
+     * This method corresponds to the convertFormat operation defined
+     * in IEEE 754.
+     *
+     * @param  v a {@link BigDecimal}
+     */
+    public static Float16 valueOf(BigDecimal v) {
+        return v.float16Value();
+    }
+
     //    /**
```
Comment (jddarcy) [inline src/java.base/share/classes/java/lang/Float16.java:437]: Please remove this comment block before pushing.

Comment (jddarcy) [reply src/java.base/share/classes/java/lang/Float16.java:437]: And the similar comment around line 90.

**Review by jddarcy — APPROVED** [review]

File: src/java.base/share/classes/java/math/BigDecimal.java [hunk 6]
```diff
@@ -3776,6 +3780,100 @@ public byte byteValueExact() {
...
+         * We have
+         *      2^12 = 2^{P+1} <= i < 2^{P+5} = 2^16
+         * Contrary to the double and float cases, where we use long and int, resp.,
+         * here we cannot simply declare i as short, because P + 5 < Short.SIZE
+         * fails to hold.
+         * Using int is safe, though.
+         *
+         * Further, as Math.scalb(Float16) does not exists, we fall back to
+         * Math.scalb(double).
+         */
```
Comment (jatin-bhateja) [inline src/java.base/share/classes/java/math/BigDecimal.java:3864]: FTR @rgiulietti , Thanks for the pointers, comments looks out of context unless read in light of https://github.com/openjdk/valhalla/blob/lworld/src/java.base/share/classes/java/math/BigDecimal.java#L4066

We should add a reference to fullDoubleValue just before your explanation.

**Review by jatin-bhateja — APPROVED** [review]

**Pull request closed without merging**